# Note: this file assumes that the gencommand_proj4_data.tar.gz file has been extracted into this directory

# Setup: Building a BowTie index in athal dir
cd gencommand_proj4
mkdir athal && bowtie2-build athal_chr.fa athal/athal


# Problem 1-5 prerequisites
## ALign both RNA seq datasets to reference genome via tophat
WORKDIR=Tophat
BWT2IDX=athal/athal

mkdir -p $WORKDIR/Day8
mkdir -p $WORKDIR/Day16

tophat -o $WORKDIR/Day8 -p 10 $BWT2IDX $DATA_DIR/Day8.fastq
tophat -o $WORKDIR/Day16 -p 10 $BWT2IDX $DATA_DIR/Day16.fastq

# Problem 1/2
## How many alignments were produced for Day8 and Day16 RNA seq datasets resp.?

### Using samtools to extract the number of alignments. This data is located in column 7
ALIGNMENTS_DAY_8=$(samtools view Tophat/Day8/accepted_hits.bam | cut -f7 | wc -l)
echo "ALIGNMENTS_DAY_8: $ALIGNMENTS_DAY_8"
## >> 63845
ALIGNMENTS_DAY_16=$(samtools view Tophat/Day16/accepted_hits.bam | cut -f7 | wc -l)
echo "ALIGNMENTS_DAY_16: $ALIGNMENTS_DAY_16"
## >> 58398

# Problem 3/4
## How many reads were mapped in each set
head Tophat/Day8/align_summary.txt
# >> 63489 reads mapped
head Tophat/Day16/align_summary.txt
# >> 57951 reads mapped

# Problem 5/6
## How many reads uniquely aligned in each set
## Subtract the (reads mapped) - (those with multiple alignments)
head Tophat/Day8/align_summary.txt
# 63133 reads uniquely aligned
head Tophat/Day16/align_summary.txt
# 57504 reads uniquely aligned

# Problem 7/8
## How many spliced alignments?
SPLICED_ALIGNMENTS_DAY_8=$(samtools view Tophat/Day8/accepted_hits.bam | cut -f6 | grep "N" | wc -l)
echo "SPLICED_ALIGNMENTS_DAY_8: $SPLICED_ALIGNMENTS_DAY_8"
# >> 8596
SPLICED_ALIGNMENTS_DAY_16=$(samtools view Tophat/Day16/accepted_hits.bam | cut -f6 | grep "N" |  wc -l)
echo "SPLICED_ALIGNMENTS_DAY_16: $SPLICED_ALIGNMENTS_DAY_16"
# >> 10695

# Problem 9/10
## How many reads were left unmapped from each set?
NUM_UNMAPPED_READS_DAY_8=$(expr 63573 - 63489)
echo "NUM_UNMAPPED_READS_DAY_8: $NUM_UNMAPPED_READS_DAY_8"
# >> 84
NUM_UNMAPPED_READS_DAY_16=$(expr 57985 - 57951)
echo "NUM_UNMAPPED_READS_DAY_16: $NUM_UNMAPPED_READS_DAY_16"
# >> 34

# Problem 11-20 prerequisites
## Assemble aligned RNA-seq reads into genes and transcripts using cufflinks

THDIR=Tophat
WORKDIR=Cufflinks

mkdir -p $WORKDIR/Day8
mkdir -p $WORKDIR/Day16

cufflinks -p 8 $THDIR/Day8/accepted_hits.bam -o $WORKDIR/Day8
cufflinks -p 8 $THDIR/Day16/accepted_hits.bam -o $WORKDIR/Day16

# Problem 11/12
## How many genes were generated by cufflinks
NUM_GENES_DAY_8=$(cat Cufflinks/Day8/transcripts.gtf | cut -f9 | cut -d ";" -f1 | sort -u | wc -l)
echo "NUM_GENES_DAY_8: $NUM_GENES_DAY_8"
## >> 186
NUM_GENES_DAY_16=$(cat Cufflinks/Day16/transcripts.gtf | cut -f9 | cut -d ";" -f1 | sort -u | wc -l)
echo "NUM_GENES_DAY_16: $NUM_GENES_DAY_16"
## >> 80

# Problem 13/14
## How many transcripts were reported
NUM_TRANSCRIPTS_DAY_8=$(cat Cufflinks/Day8/transcripts.gtf | cut -f9 | cut -d ";" -f2 | sort -u | wc -l)
echo "NUM_TRANSCRIPTS_DAY_8: $NUM_TRANSCRIPTS_DAY_8"
## >> 192
NUM_TRANSCRIPTS_DAY_16=$(cat Cufflinks/Day16/transcripts.gtf | cut -f9 | cut -d ";" -f2 | sort -u | wc -l)
echo "NUM_TRANSCRIPTS_DAY_16: $NUM_TRANSCRIPTS_DAY_16"
## >> 92

# Problem 15/16
## How many single transcripts genes were reported
NUM_SINGLE_TRANS_GENES_DAY_8=$(cat Cufflinks/Day8/transcripts.gtf | cut -f9 | cut -d ";" -f1,2 | uniq | cut -d ";" -f1 | uniq -c | awk '$1 == 1'| wc -l)
echo "NUM_SINGLE_TRANS_GENES_DAY_8: $NUM_SINGLE_TRANS_GENES_DAY_8"
## >> 180
NUM_SINGLE_TRANS_GENES_DAY_16=$(cat Cufflinks/Day16/transcripts.gtf | cut -f9 | cut -d ";" -f1,2 | uniq | cut -d ";" -f1 | uniq -c | awk '$1 == 1'| wc -l)
echo "NUM_SINGLE_TRANS_GENES_DAY_16: $NUM_SINGLE_TRANS_GENES_DAY_16"
## >> 69

# Problem 17/18
## How many single exon transcripts are there
SINGLE_EXON_TRANSCRIPTS_DAY_8=$(cat Cufflinks/Day8/transcripts.gtf | cut -f9  | grep "exon_number" | cut -d ";" -f2,3 | cut -d ";" -f1 | uniq -c | awk '$1 == 1' | wc -l)
echo "SINGLE_EXON_TRANSCRIPTS_DAY_8: $SINGLE_EXON_TRANSCRIPTS_DAY_8"
## >> 119
SINGLE_EXON_TRANSCRIPTS_DAY_16=$(cat Cufflinks/Day16/transcripts.gtf | cut -f9  | grep "exon_number" | cut -d ";" -f2,3 | cut -d ";" -f1 | uniq -c | awk '$1 == 1' | wc -l)
echo "SINGLE_EXON_TRANSCRIPTS_DAY_16: $SINGLE_EXON_TRANSCRIPTS_DAY_16"
## >> 24


# Problem 19/20
## How many multi-exon transcripts are there
MULTI_EXON_TRANSCRIPTS_DAY_8=$(cat Cufflinks/Day8/transcripts.gtf | cut -f9  | grep "exon_number" | cut -d ";" -f2,3 | cut -d ";" -f1 | uniq -c | awk '$1 > 1' | wc -l)
echo "MULTI_EXON_TRANSCRIPTS_DAY_8: $MULTI_EXON_TRANSCRIPTS_DAY_8"
## >> 73
MULTI_EXON_TRANSCRIPTS_DAY_16=$(cat Cufflinks/Day16/transcripts.gtf | cut -f9  | grep "exon_number" | cut -d ";" -f2,3 | cut -d ";" -f1 | uniq -c | awk '$1 > 1' | wc -l)
echo "MULTI_EXON_TRANSCRIPTS_DAY_16: $MULTI_EXON_TRANSCRIPTS_DAY_16"
## >> 68


# Problem 21-30 prerequisites
## ALign both RNA seq datasets to reference genome via tophat
DATA_DIR=Cufflinks
WORKDIR=Cuffcompare

mkdir -p $WORKDIR/Day8
mkdir -p $WORKDIR/Day16

cuffcompare -r athal_genes.gtf -R $DATA_DIR/Day8/transcripts.gtf -o $WORKDIR/Day8
cuffcompare -r ./athal_genes.gtf -R $DATA_DIR/Day16/transcripts.gtf -o $WORKDIR/Day16

# Problem 21/22
## How many cufflinks transcripts fully reconstruct annotation transcripts in each sample
## Grep for full matches ("=")
NUM_CUFFLINK_TRANS_RECONSTRUCT_ANNOT_TRANS_DAY_8=$(cat Cufflinks/Day8/Day8.transcripts.gtf.tmap | cut -f3 | grep "=" | wc -l)
echo "NUM_CUFFLINK_TRANS_RECONSTRUCT_ANNOT_TRANS_DAY_8: $NUM_CUFFLINK_TRANS_RECONSTRUCT_ANNOT_TRANS_DAY_8"
## >> 16
NUM_CUFFLINK_TRANS_RECONSTRUCT_ANNOT_TRANS_DAY_16=$(cat Cufflinks/Day16/Day16.transcripts.gtf.tmap | cut -f3 | grep "=" | wc -l)
echo "NUM_CUFFLINK_TRANS_RECONSTRUCT_ANNOT_TRANS_DAY_16: $NUM_CUFFLINK_TRANS_RECONSTRUCT_ANNOT_TRANS_DAY_16"
## >> 36

# Problem 23/24
## How many splice variants does the gene AT4G20240 have in the two samples?
NUM_SPLICE_VARIANTS_AT4G20240_DAY_8=$(cat Cufflinks/Day8/Day8.transcripts.gtf.tmap | grep AT4G20240 | wc -l)
echo "NUM_SPLICE_VARIANTS_AT4G20240_DAY_8: $NUM_SPLICE_VARIANTS_AT4G20240_DAY_8"
## >> 2
NUM_SPLICE_VARIANTS_AT4G20240_DAY_16=$(cat Cufflinks/Day16/Day16.transcripts.gtf.tmap | grep AT4G20240 | wc -l)
echo "NUM_SPLICE_VARIANTS_AT4G20240_DAY_16: $NUM_SPLICE_VARIANTS_AT4G20240_DAY_16"
## >> 0

# Problem 25/26
## How many splice variants are partial reconstructions of a reference transcript "contained"?
## Grep for contained matched ("c")
NUM_SPLICE_VARIANTS_CONTAINED_DAY_8=$(cat Cufflinks/Day8/Day8.transcripts.gtf.tmap | cut -f3 | grep "c" | wc -l)
echo "NUM_SPLICE_VARIANTS_CONTAINED_DAY_8: $NUM_SPLICE_VARIANTS_CONTAINED_DAY_8"
## >> 133
NUM_SPLICE_VARIANTS_CONTAINED_DAY_16=$(cat Cufflinks/Day16/Day16.transcripts.gtf.tmap | cut -f3 | grep "c" | wc -l)
echo "NUM_SPLICE_VARIANTS_CONTAINED_DAY_16: $NUM_SPLICE_VARIANTS_CONTAINED_DAY_16"
## >> 21

# Problem 27/28
## How many cufflinks transcripts are novel splice variants of reference genes?
## Grep for contained matched ("j")
NUM_NOVEL_SPLICE_VARIANTS_DAY_8=$(cat Cufflinks/Day8/Day8.transcripts.gtf.tmap | cut -f3 | grep "j" | wc -l)
echo "NUM_NOVEL_SPLICE_VARIANTS_DAY_8: $NUM_NOVEL_SPLICE_VARIANTS_DAY_8"
## >> 14
NUM_NOVEL_SPLICE_VARIANTS_DAY_16=$(cat Cufflinks/Day16/Day16.transcripts.gtf.tmap | cut -f3 | grep "j" | wc -l)
echo "NUM_NOVEL_SPLICE_VARIANTS_DAY_16: $NUM_NOVEL_SPLICE_VARIANTS_DAY_16"
## >> 22

# Problem 29/30
## How many cufflinks transcripts were formed in the introns of reference genes?
## Grep for contained matched ("i")
NUM_INTRON_SPLICE_VARIANTS_DAY_8=$(cat Cufflinks/Day8/Day8.transcripts.gtf.tmap | cut -f3 | grep "i" | wc -l)
echo "NUM_INTRON_SPLICE_VARIANTS_DAY_8: $NUM_INTRON_SPLICE_VARIANTS_DAY_8"
## >> 4
NUM_INTRON_SPLICE_VARIANTS_DAY_16=$(cat Cufflinks/Day16/Day16.transcripts.gtf.tmap | cut -f3 | grep "i" | wc -l)
echo "NUM_INTRON_SPLICE_VARIANTS_DAY_16: $NUM_INTRON_SPLICE_VARIANTS_DAY_16"
## >> 1

# Problem 31-35 prerequisites
## ALign both RNA seq datasets to reference genome via tophat
THDIR=Tophat
ANNOT=$(pwd/)athal_genes.gtf
GTFs=Cuffmerge/GTFs.txt

rm -rf Cuffmerge
mkdir -p Cuffmerge
touch Cuffmerge/GTFs.txt

echo "$(pwd)/Cufflinks/Day8/transcripts.gtf" >> $GTFs
echo "$(pwd)/Cufflinks/Day16/transcripts.gtf" >> $GTFs

# Cuffmerge to merge and reconcile two sets of cufflinks transcripts
cuffmerge -g $ANNOT -p 8 -o Cuffmerge $GTFs

# Cuffdiff to perform DEA
cuffdiff -o Cuffdiff -p 10 Cuffmerge/merged.gtf $THDIR/Day8/accepted_hits.bam $THDIR/Day16/accepted_hits.bam

# Problem 31
## How many genes (loci) were reported in the merged gtf file
NUM_MERGED_LOCI=$(cat Cuffmerge/merged.gtf | cut -f9 | cut -d ";" -f1 | uniq | wc -l)
echo "NUM_MERGED_LOCI: $NUM_MERGED_LOCI"
# >> 129

# Problem 32
## How many transcripts
NUM_MERGED_TRANSCRIPTS=$(cat Cuffmerge/merged.gtf | cut -f9 | cut -d ";" -f2 | uniq | wc -l)
echo "NUM_MERGED_TRANSCRIPTS: $NUM_MERGED_TRANSCRIPTS"
# >> 200

# Problem 33
## Num total genes included in the gene expression report from cuffdiff
NUM_TOTAL_CUFFDIFF_GENES=$(cat Cuffdiff/gene_exp.diff | cut -f2 | uniq | wc -l)
echo "NUM_TOTAL_CUFFDIFF_GENES: $NUM_TOTAL_CUFFDIFF_GENES"
# >> 129

# Problem 34
## How many genes are (significantly) differentially expressed
NUM_DIFF_EXPR_GENES=$(cat Cuffdiff/gene_exp.diff | cut -f14 | grep "yes" | wc -l)
echo "NUM_DIFF_EXPR_GENES: $NUM_DIFF_EXPR_GENES"
# >> 4

# Problem 35
## How many transcripts were differentially expressed between the two samples
NUM_DIFF_EXPR_TRANS=$(cat Cuffdiff/isoform_exp.diff | cut -f14 | grep "yes" | wc -l)
echo "NUM_DIFF_EXPR_TRANS: $NUM_DIFF_EXPR_TRANS"
# >> 5

